contrib/educational_decoder/zstd_decompress.c

```diff
@@ -0,0 +1,2168 @@
+
+    // decode window size
+    if (!single_segment_flag) {
+        if (src_len < header_size + 1) {
+            INP_SIZE();
+        }
+
+        // Use the algorithm from the specification to compute window size
+        // https://github.com/facebook/zstd/blob/dev/doc/zstd_compression_format.md#window_descriptor
+        u8 window_descriptor = src[header_size];
+        u8 exponent = window_descriptor >> 3;
+        u8 mantissa = window_descriptor & 7;
+
+        size_t window_base = (size_t)1 << (10 + exponent);
+        size_t window_add = (window_base / 8) * mantissa;
+        header->window_size = window_base + window_add;
+
+        header_size++;
+    }
+
+    // decode dictionary id if it exists
+    if (dictionary_id_flag) {
+        const int bytes_array[] = {0, 1, 2, 4};
+        const int bytes = bytes_array[dictionary_id_flag];
+
+        if (src_len < header_size + bytes) {
+            INP_SIZE();
+        }
+
+        header->dictionary_id = read_bits_LE(src + header_size, bytes * 8, 0);
+
+        header_size += bytes;
+    } else {
+        header->dictionary_id = 0;
+    }
+
+    // decode frame content size if it exists
+    if (single_segment_flag || frame_content_size_flag) {
+        // if frame_content_size_flag == 0 but single_segment_flag is set, we
+        // still have a 1 byte field
+        const int bytes_array[] = {1, 2, 4, 8};
+        const int bytes = bytes_array[frame_content_size_flag];
+
+        if (src_len < header_size + bytes) {
+            INP_SIZE();
+        }
+
+        header->frame_content_size =
+            read_bits_LE(src + header_size, bytes * 8, 0);
+        if (bytes == 2) {
+            header->frame_content_size += 256;
+        }
+
+        header_size += bytes;
+    } else {
+        header->frame_content_size = 0;
+    }
+
+    if (single_segment_flag) {
+        // in this case the effective window size is frame_content_size this
+        // impacts sequence decoding as we need to determine whether to fall
+        // back to the dictionary or not on large offsets
+        header->window_size = header->frame_content_size;
+    }
+
+    header->header_size = header_size;
+}
+
+/// A dictionary acts as initializing values for the frame context before
+/// decompression, so we implement it by applying it's predetermined
+/// tables and content to the context before beginning decompression
+static void frame_context_apply_dict(frame_context_t *const ctx,
+                                     const dictionary_t *const dict) {
+    // If the content pointer is NULL then it must be an empty dict
+    if (!dict || !dict->content)
+        return;
+
+    if (ctx->header.dictionary_id == 0 && dict->dictionary_id != 0) {
+        // The dictionary is unneeded, and shouldn't be used as it may interfere
+        // with the default offset history
+        return;
+    }
+
+    // If the dictionary id is 0, it doesn't matter if we provide the wrong raw
```

Comment:
So the only case to do any early returns here is if the header has a non-zero id and the provided dictionary doesn't match it, right?